### ISSUE TITLE
Nested unsafeRuns cause deadlock & lost exceptions

### ISSUE THREAD

**Issue by kaishh** [scalaz/scalaz-zio#277]
Example:

```scala
import java.util.concurrent._

import scalaz.zio.{IO, RTS}

object NestedUnsafeRunTest extends App {
  val rts = new RTS {
    override val threadPool: ExecutorService = {
      val executor = new ThreadPoolExecutor(
        8, 64,
        60, TimeUnit.SECONDS,
        new LinkedBlockingDeque[Runnable](64),
        Executors.defaultThreadFactory())

      executor
    }
  }

  rts.unsafeRunSync {
    val nestedRun = (1 to 300).foldLeft (IO.unit) {
      (io, i) => IO.sync(rts.unsafeRun {
        println(s"nested run $i")
        io
      }).run.void
    }

    nestedRun
  }
}
```

Expected:

`RejectedExecutionException` due to exhausted thread pool.

Actual:

Deadlock. `RejectedExecutionException` is received at some point and is continuously rethrown, but it doesn't terminate.

```
nested run 300
nested run 299
nested run 298
nested run 297
nested run 296
nested run 295
nested run 294
nested run 293
```

**Comment by pshirshov** [comment]
Uhu, uhu. We almost got this stuff in real prod 😠

**Comment by pshirshov** [comment]
> Expected: RejectedExecutionException due to exhausted thread pool.

`Terminated(_:RejectedExecutionException)` to be more precise

**Comment by regiskuckaertz** [comment]
That may be a silly question, but in what situation would you want to do that?

**Comment by pshirshov** [comment]
Our teammate needed a cache of limited size with soft references so he wrote this chunk:

```
itemRepository: ItemRepository[F]
  
  private val cacheLoader: CacheLoader[Id, CachedValue] = new CacheLoader[Id, CachedValue]() {
    override def load(app: Id): CachedValue = {
      BIORunner[F].unsafeRun {
        enumerate(app).map(r => CachedValue(r))
      }
    }
  }

  private val cache: LoadingCache[Id, CachedValue] = CacheBuilder
    .newBuilder()
    .maximumSize(itemsCacheConfig.size)
    .expireAfterWrite(itemsCacheConfig.ttl, TimeUnit.SECONDS)
    .softValues()
    .build(cacheLoader)
```

Actually, he hand't much choice. This case is kinda crappy but still legal.

**Comment by jdegoes** [comment]
Thanks for reporting! This looks like it should work to me.

**Comment by pshirshov** [comment]
Sorry for spamming, just left it here for future reference.

**Comment by NeQuissimus** [comment]
@pshirshov I folded up the thread dump for you ;)

**Comment by pshirshov** [comment]
Ah, thanks. 

<details> Never knew about this feature. </details>

**Comment by regiskuckaertz** [comment]
@kaishh Try by removing the `.run` as it reifies the error in the value channel. I'm able to run your program consistently with that modification.

**Comment by regiskuckaertz** [comment]
Introducing a delay will block though.
```scala
val nestedRun = (1 to 300).foldLeft (IO.unit) {
  (io, i) => IO.sync(rts.unsafeRun {
    println(s"nested run $i")
    io.delay(1.second)
  })
}
```
I'll keep on investigating.

**Comment by kaishh** [comment]
@regiskuckaertz the point of the test is to check if forking is consistent in presence of intermediate blockinng unsafeRuns, I.e. .run or fork.join should make it exhaust the pool, but not diverge.

**Comment by pshirshov** [comment]
From my point of view rejection must be treated specifically and lead to `Terminated()` exit result.

**Comment by regiskuckaertz** [comment]
There isn't an opportunity to exhaust the thread pool or throw an rejected exception, because each task blocks on its child completing (with `run`):
```scala
IO.sync(rts.unsafeRun {
  println(s"nested run 300")
  IO.sync(rts.unsafeRun {
    println(s"nested run 299")
    IO.sync(rts.unsafeRun {
      println(s"nested run 298")
      ...
        IO.sync(rts.unsafeRun {
          println(s"nested run 1")
          IO.unit
        }).run.void
      ...
    }).run.void
  }).run.void
}).run.void
```
After `corePoolSize` tasks have been submitted, the next one is added to the queue and this is where the program stops, because no thread will be created until the queue is full.

> RejectedExecutionException is received at some point and is continuously rethrown

I haven't experienced that on my machine; was it with the exact same code?

Replacing `run` with `fork` works as expected.

**Comment by pshirshov** [comment]
> After corePoolSize tasks have been submitted, the next one is added to the queue and this is where the program stops, because no thread will be created until the queue is full.

Yup, it should _stop_ in case the queue backing thread pool rejected to take new task, not got stuck.

> I haven't experienced that on my machine; was it with the exact same code?

Yes, exactly the same and it reproduces always. We are on jdk10/jdk11. If you wanna we may check it on 8 as well.

**Comment by pshirshov** [comment]
@jdegoes I've tried to check your theory with the following dirty snippet:

```
  def defaultHandler: List[Throwable] => IO[Nothing, Unit] =
    (ts: List[Throwable]) => {
      ts.foreach(_.printStackTrace())
      System.exit(1)
      IO.sync(ts.foreach(_.printStackTrace()))
    }
```

But the test hanged, didn't exit.

<details>
<summary>The code is still the same</summary>

```
package scalaz.zio.interop

import java.util.concurrent._

import scalaz.zio.{ IO, RTS }

object NestedUnsafeRunTest {
  val rts = new RTS {
    override val threadPool: ExecutorService = {
      val executor = new ThreadPoolExecutor(8,
                                            64,
                                            60,
                                            TimeUnit.SECONDS,
                                            new LinkedBlockingDeque[Runnable](64),
                                            Executors.defaultThreadFactory())

      executor
    }
  }

  def main(args: Array[String]): Unit = {
    val _ = rts.unsafeRunSync {

      val nestedRun = (1 to 300).foldLeft(IO.unit) { (io, i) =>
        IO.sync(
            rts.unsafeRun {
              println(s"nested run $i")
              io
            }
          )
          .run
          .void

      }
      nestedRun
    }
  }

}
```

</details>

The behaviour is the same on jdk8 so it seems to be consistent.

**Comment by kaishh** [comment]
@regiskuckaertz 
Sorry! The exception is indeed never thrown with `LinkedBlockingQueue`. I've tried different queues before submitting the report and did not recheck. For `SynchronousQueue` setting a breakpoint on `RejectedExecutionException` does show me unending throws.

@pshirshov 
The fork happens in `RTS.register`:
```
case ExitResult.Terminated(ts) =>
  rts.submit(rts.unsafeRun(unhandled(ts)))
  cb(ExitResult.Terminated(ts))
```

Overriding defaultHandler can't help it.

**Comment by regiskuckaertz** [comment]
Sorry, I meant `hang`, not `stop` indeed; lost in translation 😓 From my understanding of how a thread pool works, threads above `corePoolSize` will only be created when the queue is full. This example is kind of the perfect storm since:

- previously submitted task will only complete when the subsequent ones have
- a task is submitted by the previously submitted task
- after `corePoolSize` have been submitted, the next one is put on the queue but is never started, since it is itself responsible for submitting the next task
- since the queue only ever contains a single task, the program hangs

This is how I explain the behaviour you observed, and the fact it worked when I changed `run` to `fork` seemed to confirm my suspicion (all 300 tasks ran, despite changing the queue size to 5). I also could see rejected exceptions thrown and handled by the default handler after playing around with the queue size and submitting all tasks in parallel.

**Comment by regiskuckaertz** [comment]
@kaishh ah, ok gotcha 🤗 Thanks for looking into it, I'll investigate using a `SynchronousQueue` instead 👍

**Comment by pshirshov** [comment]
Same problem with 

```
        case ExitResult.Terminated(ts) =>
          System.exit(1)
          rts.submit(rts.unsafeRun(unhandled(ts)))
          cb(ExitResult.Terminated(ts))
```

On `SynchronousQueue`.

All the threads stuck with ```
"pool-2-thread-42" #52 prio=5 os_prio=31 tid=0x00007f8cb88a7000 nid=0x7903 in Object.wait() [0x000070000b6f6000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c063d170> (a java.util.concurrent.atomic.AtomicReference)
	at java.lang.Object.wait(Object.java:502)
	at scalaz.zio.RTS$FiberContext.runSync(RTS.scala:209)
	- locked <0x00000006c063d170> (a java.util.concurrent.atomic.AtomicReference)
	at scalaz.zio.RTS.unsafeRunSync(RTS.scala:40)
	at scalaz.zio.RTS.unsafeRunSync$(RTS.scala:37)
	at scalaz.zio.interop.NestedUnsafeRunTest$$anon$1.unsafeRunSync(NTT.scala:8)
	at scalaz.zio.RTS.unsafeRun(RTS.scala:21)
	at scalaz.zio.RTS.unsafeRun$(RTS.scala:21)
	at scalaz.zio.interop.NestedUnsafeRunTest$$anon$1.unsafeRun(NTT.scala:8)
	at scalaz.zio.interop.NestedUnsafeRunTest$.$anonfun$main$2(NTT.scala:30)
	at scalaz.zio.interop.NestedUnsafeRunTest$$$Lambda$11/2036958521.apply$mcV$sp(Unknown Source)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scalaz.zio.RTS$FiberContext.evaluate(RTS.scala:398)
	at scalaz.zio.RTS$FiberContext.$anonfun$fork$1(RTS.scala:618)
	at scalaz.zio.RTS$FiberContext$$Lambda$24/510113906.apply$mcV$sp(Unknown Source)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scalaz.zio.RTS$$anon$1.run(RTS.scala:81)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)```

 I didn't dig deeper.

**Comment by regiskuckaertz** [comment]
@jdegoes Yes. _When_ rejection occurs, it manifests itself as an exception treated by the error handler (which can be provided via `fork0`) and will also lead to a `Terminated(RejectedExecutionException)`. That happens [here](https://github.com/scalaz/scalaz-zio/blob/master/core/jvm/src/main/scala/scalaz/zio/RTS.scala#L913-L915), and `unsafeRun` will receive it [here](https://github.com/scalaz/scalaz-zio/blob/master/core/jvm/src/main/scala/scalaz/zio/RTS.scala#L196-L212).

However with this program it is impossible to trigger a rejection, since `run` observes the forked fiber, i.e. it waits until the fiber terminates. This is essentially what it does:
```
threadPool.submit(new Runnable {
  threadPool.submit(new Runnable {
    threadPool.submit(new Runnable {
      ... // 300 layers deep
    })
  })
})
```
With `run`, each task waits for its child to be done (that's the meaning of `run`), that is what brings the program to a hang state, since after `corePoolSize + 1` tasks have been submitted, there are `corePoolSize` threads in a waiting state and one in the queue that will never get started, [since](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ThreadPoolExecutor.html):

> a new thread will be created only if the queue is full

The task in the queue is never started and so no other task will ever get submitted. Using `fork` instead does not change the shape of the program, but now each task creates a new one and moves on with its own work. Eventually, it finishes and so the task that is in the queue can be started on an idle thread of the core pool, submit a new task, and so on until all tasks have run.

However using `fork` alone is not ideal, since execution of the fibers will leak out of each IO action. You need to join them all. Just to prove that ZIO properly catches rejections, this version:
```scala
  val result = rts.unsafeRunSync {
    val nestedRun = IO.traverse((1 to 300)) {
      i => IO.sync(rts.unsafeRun {
        IO.sync(println(s"nested run $i"))
      }).fork
    }

    nestedRun.flatMap(Fiber.joinAll(_))
  }

  println(s"result is $result")
```
produces (changing the queue size to 1 and running it a couple of times):
```
result is Terminated(List(java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.FutureTask@28d6c6f3 rejected from java.util.concurrent.ThreadPoolExecutor@326c2a09[Running, pool size = 64, active threads = 1, queued tasks = 1, completed tasks = 119]))
```
Arguably, the nesting has disappeared. Getting it back is tricky and I haven't found a way yet.

**Comment by jdegoes** [comment]
So the scenario is essentially:

 - Task 1: Start, wait on Task 2
 - Task 2: Start, wait on Task 3
 - Task 3: Start, wait on Task 4
 - ...
 - Task N: Start, wait on Task N + 1.
 - Task N + 1: Hold (in thread pool queue)

where `N` = `maximumPoolSize`. 

Separately, we have a hypothetical (?) problem of infinite errors due to `RejectedExecutionException` (which should surely be looked at and fixed somehow).

The core problem happens because all threads are waiting for a task in the queue to finish; but the task in the queue will never finish, because it will never be started, because all threads are busy waiting for it to finish.

It may be possible to detect such a situation. I think that's the best we can hope for, unless I'm missing something.

**Comment by regiskuckaertz** [comment]
> Separately, we have a hypothetical (?) problem of infinite errors due to RejectedExecutionException 

Haven't yet had the time to look into that.

> It may be possible to detect such a situation

Via a separate `IO` interpreter? Something taking IO into a graph data structure that can be analysed for degenerate cases like this one.

**Comment by pshirshov** [comment]
A livelock on pool exhaustion is, um, expected behavior. It's possible to detect it with a heuristic but you can't solve all class of these problems, they are  fundamental.

What needs to be fixed is that `RejectedExecutionException` livelock occurring when pool stops accepting new tasks. This should be enough.

**Comment by jdegoes** [comment]
> What needs to be fixed is that RejectedExecutionException livelock occurring when pool stops accepting new tasks. This should be enough.

We could solve this in several ways:

1. A separate dedicated thread with an unbounded queue for dispatching errors. This would solve this problem specifically, because it would mean that reporting an unhandled error such as `RejectedExecutionException` would not cause further errors of the same type.
2. Not forking uncaught error handlers (which is the current behavior); but rather assuming they will be short, finite computations that can execute on the current thread. The stack size could grow in such a situation and would require careful monitoring.
3. Detecting and ignoring defects from unhandled error handlers. This would solve the problem more generally, as potentially any top-level unhandled error handler that does, say, `throw new Error`, would have a similar problem of triggering an infinite cascade of errors (to itself).
4. Making sure any exception in an uncaught error handler is handled by the parent, and introducing a new parent that cannot be overridden which has infallible behavior. This would not solve the current problem but would solve all cases of handlers that throw exceptions which are then handled by themselves.

Possibly (1) and (4) together are the best combination.

**Comment by pshirshov** [comment]
Tbh I would go with (2). This would require the minimal amount of new code. In case you create a dedicate thread you may need to start taking care of it's termination for example. Of course you would make it daemon but someone may need to instantiate multiple `RTSses` during app lifecycle, so these threads may become a problem in case implemented without taking care of all the potential usecases.

**Comment by regiskuckaertz** [comment]
It seems a good heuristic would be to keep track of (1) the "depth" of a fiber and (2) whether the parent fiber is observing it; when `depth == corePoolSize && ∀ f ∈ ancestors(this). observed(f)`, the RTS could reject any new task. Easier said than done 😄

**Comment by jdegoes** [comment]
Dispatching unhandled errors now happens outside the runtime system and with an ordinary effectful method defined on `Platform`. Closing this issue, please re-open if there is more work we need to do.
